lambda_builders/actions/python_pip/DESIGN.md

```diff
@@ -0,0 +1,149 @@
+## Python - PIP Lambda Builder
+
+### Scope
+
+This package is an effort to port the Chalice packager to a library that can
+be used to handle the dependency resilution portion of packaging Python code
+for use in AWS Lambda. The scope for this builder is to take an existing
+directory containing customer code, and a top-levle `requirements.txt` file
+specifying third party depedencies. The builder will examine the dependencies
+and use pip to build and include the dependencies in the customer code bundle
+in a way that makes them importable.
+
+### Challenges
+
+Python is particularly difficult to package for other platforms because it is
+heavily coupled to your local environment. This is because `setup.py` the
+"config" to describe an install is just a python file, which means authors can
+run code to check things about your current system, and then use that to make
+decisions about how to install the package. These decisions are obviously not
+valid once we move the built package to a different platform.
+
+Python packaging also has the concept of a wheel, which is more of a self
+contained package that can be trivially installed into a python environment
+and does not execute any code. Our goal is to build up a set of these that
+are known to be compatible with AWS Lambda.
+
+### Interface
+
+The top level interface will consist of a single function with the signature:
+
+```python
+def build_depenencies(source_path, config=None):
+    """Builds a source directory's dependencies into the directory.
+
+	:type source_path: str
+	:param source_path: The source directory that needs to have dependencies
+		built for it. It is assumed that this customer code is correct and
+		would function if it had all of its dependencies installed alongside
+		it. It is assumed that the dependencies are specified in a top level
+		``requirements.txt`` file. This file can be removed before deploying
+		to lambda, but the this function will not modify any existing files
+		in this directory.
+
+	:type config: :class:`lambda_builders.actions.python_pip.config`
+	:param config: A configuration object which contains additional config
+		to control how the source directories dependencies are installed.
+		This may include concepts such as skips where a dependency is
+		already prepared and the builder need not waste time compiling that
+		dependency.
+	"""
+```
+
+There are a couple ways error reporting can be done.
+
+1) We can raise a python error and have the caller catch and deal with
+   it at the end. The error may not be significant, for example if the user
+   is already aware of the problem and has a post-build hook that fixes the
+   bundle with a custom dependency injection. So this exception would be thrown
+   at the end and not interrupt the rest of the dependency building.
+2) Return None or a list of aggregated error objects that can be inspected by
+   the caller.
+3) Use the UI object. The builder as it is written right now has a notion of a
```

Comment:
I like the idea of a UI object. This allows us to abstract functionality like "write with colors" etc within the object. We could also redirect the data to a file transparently. Feel free to make assumptions about this object for now. We can implement a real UI object separately.

Reply:
Cool. For now Whatever wrapper around this low level builder can create and inject it.